mlflow/langchain/chat_utils.py

```diff
@@ -0,0 +1,294 @@
+
+
+# NB: Even though _ChatRequest is only referenced in one method within this module
+# (as of 12/27/2023), it must be defined at the module level for compatibility with
+# pydantic < 2
+class _ChatRequest(pydantic.BaseModel, extra="forbid"):
+    messages: List[_ChatMessage]
+
+
+class _ChatChoice(pydantic.BaseModel, extra="forbid"):
+    index: int
+    message: _ChatMessage = None
+    finish_reason: Optional[str] = None
+
+    @staticmethod
+    def get_schema():
+        return Schema(
+            [
+                ColSpec(DataType.integer, "index"),
+                ColSpec(_ChatMessage.get_schema(), "message", required=False),
+                ColSpec(DataType.string, "finish_reason", required=False),
+            ]
+        )
+
+
+class _ChatChoiceDelta(pydantic.BaseModel):
+    index: int
+    finish_reason: Optional[str] = None
+    delta: _ChatDeltaMessage
+
+
+class _ChatUsage(pydantic.BaseModel, extra="forbid"):
+    prompt_tokens: Optional[int] = None
+    completion_tokens: Optional[int] = None
+    total_tokens: Optional[int] = None
+
+    @staticmethod
+    def get_schema():
+        return Schema(
+            [
+                ColSpec(DataType.integer, "prompt_tokens", required=False),
+                ColSpec(DataType.integer, "completion_tokens", required=False),
+                ColSpec(DataType.integer, "total_tokens", required=False),
+            ]
+        )
+
+
+class _ChatResponse(pydantic.BaseModel, extra="forbid"):
+    id: Optional[str] = None
+    object: Literal["chat.completion"] = "chat.completion"
+    created: int
+    # Make the model field optional since we may not be able to get a stable model identifier
+    # for an arbitrary LangChain model
+    model: Optional[str] = None
+    choices: List[_ChatChoice]
+    usage: _ChatUsage
+
+    @staticmethod
+    def get_schema():
+        return Schema(
+            [
+                ColSpec(DataType.string, "id", required=False),
+                ColSpec(DataType.string, "object"),
+                ColSpec(DataType.integer, "created"),
+                ColSpec(DataType.string, "model", required=False),
+                ColSpec(Array(_ChatChoice.get_schema()), "choices"),
+                ColSpec(_ChatUsage.get_schema(), "usage"),
+            ]
+        )
+
+
+class _ChatChunkResponse(pydantic.BaseModel):
+    id: Optional[str] = None
+    object: Literal["chat.completion.chunk"] = "chat.completion.chunk"
+    created: int
+    # Make the model field optional since we may not be able to get a stable model identifier
+    # for an arbitrary LangChain model
+    model: Optional[str] = None
+    choices: List[_ChatChoiceDelta]
+
+
+def _try_transform_response_to_chat_format(response):
+    if isinstance(response, str):
+        message_content = response
+        message_id = None
+    elif isinstance(response, AIMessage):
+        message_content = response.content
+        message_id = getattr(response, "id", None)
+    else:
+        return response
```

Comment:
Suggestion: what about adding a quick comment for what should fall into `else`? There are a few places we do `if-elif-else`, but what kind of inputs are expected in `else` is not visible from the code at all. Alternatively, we can add a type hint for response like `Union[str, AIMessage, ...]`.

Reply:
Here response can be anything, whatever your model returns. Do we have to list them out here? I feel like as long as we make it clear that we only convert response if it's (str, AImessage) then it's enough?

Reply:
Got it, if the list is too long it's ok to leave, tho it sounds like a big source of unknown unknown....